### PR TITLE
fixing app crash on location change

### PR DESCRIPTION
**Pull request by pr4bh4sh**
Fixing crash when trying to set the location on Pie emulator

Adb command 
```bash
adb shell am start-foreground-service --user 0 -n io.appium.settings/.LocationService --es longitude -87.12733 --es latitude 15.13697
```

Crash log
```
--------- beginning of crash
08-07 11:16:02.437 E/AndroidRuntime( 6577): FATAL EXCEPTION: main
08-07 11:16:02.437 E/AndroidRuntime( 6577): Process: io.appium.settings, PID: 6577
08-07 11:16:02.437 E/AndroidRuntime( 6577): android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground()
08-07 11:16:02.437 E/AndroidRuntime( 6577):     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1775)
08-07 11:16:02.437 E/AndroidRuntime( 6577):     at android.os.Handler.dispatchMessage(Handler.java:105)
08-07 11:16:02.437 E/AndroidRuntime( 6577):     at android.os.Looper.loop(Looper.java:164)
08-07 11:16:02.437 E/AndroidRuntime( 6577):     at android.app.ActivityThread.main(ActivityThread.java:6541)
08-07 11:16:02.437 E/AndroidRuntime( 6577):     at java.lang.reflect.Method.invoke(Native Method)
08-07 11:16:02.437 E/AndroidRuntime( 6577):     at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
08-07 11:16:02.437 E/AndroidRuntime( 6577):     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
08-07 11:16:02.464 D/gralloc_ranchu( 1438): gralloc_alloc: Creating ashmem region of size 2289664
08-07 11:16:02.470 D/        ( 6921): HostConnection::get() New Host Connection established 0x7db95082f780, tid 6921
08-07 11:16:02.479 E/EGL_emulation( 1448): tid 1448: eglCreateSyncKHR(2049): error 0x3004 (EGL_BAD_ATTRIBUTE)
08-07 11:16:02.485 I/ActivityManager( 1792): Showing crash dialog for package io.appium.settings u0
08-07 11:16:02.546 D/gralloc_ranchu( 1438): gralloc_alloc: Creating ashmem region of size 1024000
08-07 11:16:02.549 I/chatty  ( 1438): uid=1000(system) allocator@2.0-s identical 1 line
08-07 11:16:02.551 D/gralloc_ranchu( 1438): gralloc_alloc: Creating ashmem region of size 1024000
08-07 11:16:02.594 D/gralloc_ranchu( 1438): gralloc_alloc: Creating ashmem region of size 770048
08-07 11:16:02.598 I/chatty  ( 1438): uid=1000(system) allocator@2.0-s identical 1 line
08-07 11:16:02.601 D/gralloc_ranchu( 1438): gralloc_alloc: Creating ashmem region of size 770048
08-07 11:16:02.620 D/EGL_emulation( 1792): eglMakeCurrent: 0x76cd36681f00: ver 2 0 (tinfo 0x76cd3253dfa0)
08-07 11:16:02.625 I/zygote64( 1792): Background concurrent copying GC freed 75006(3MB) AllocSpace objects, 23(2MB) LOS objects, 34% free, 11MB/17MB, paused 530us total 173.842ms
```

### CODE REVIEW

**Comment by pr4bh4sh** [comment]
I've tested this manually and the fix works. 
Issue to address 
1. After this fix, there will to 2 appium setting notification. And because of `scheduleLocationUpdate()` not sure at what point the service should be stopped.
2. Refactor the code

**Comment by mykola-mokhnach** [comment]
We probably need to be smarter about that. In general, Appium Settings itself does put itself into foreground if started from the main activity. Such exception might only happen if one tries to perform a broadcast without the app us running. 
So, my proposal is:
- We should check if the app is already running and should not set it to run in foreground if it is already running
- We should avoid unnecessary code duplication

**Comment by pr4bh4sh** [comment]
I agree with both points.

For my case, I'm using the only Appium settings app, not with Appium server. I'm using the settings app to set the location. Here's what I do
1. Install the app
2. Check if the service is running with `adb shell dumpsys activity services 'io.appium.settings'` if not give all permission need for setting the mock location and start the app with `adb shell am start io.appium.settings/.Settings'`
3. Check again if the appium service is running now, if yes then set the location with `adb shell am start-foreground-service --user 0 -n io.appium.settings/.LocationService --es longitude #{longitude} --es latitude #{latitude}`

The problem is even after doing all of this the setting app crashes.

The changes I made worked for me and I see 0 crashes. 
I'm not sure if other users are facing the same issue. I can create a thread on appium forum and check if someone is facing this issue.

**Comment by mykola-mokhnach** [comment]
I get your point and I think the changes are good and useful. I would be happy to merge them after refactoring.

> The problem is even after doing all of this the setting app crashes.

Yep, I saw situations when sometimes Android can kill an activity on its own, so this makes a lot of sense to me.

**Comment by pr4bh4sh** [comment]
Ok, I will refactor it.

File: app/src/main/java/io/appium/settings/LocationService.java [hunk 1]
```diff
@@ -77,9 +78,17 @@ public int onStartCommand(Intent intent, int flags, int startId) {
                 return START_NOT_STICKY;
             }
         }
+
+        // https://stackoverflow.com/a/45047542
+        // https://developer.android.com/about/versions/oreo/android-8.0-changes.html
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            Log.w(TAG, "Starting location service");
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/settings/LocationService.java]: I'd rather change the logging level to info or debug

Comment (pr4bh4sh) [reply app/src/main/java/io/appium/settings/LocationService.java]: 👍  added it to check the log, forgot to revert it. Changing it to info.

**Comment by mykola-mokhnach** [comment]
This fix has been published in v. 2.14.2

**Comment by KazuCocoa** [comment]
👍

**Pull request merged**